Automattic/wc-calypso-bridge
# PayPal: Verify method_supports refunds in API response

This branch seeks to fix https://github.com/Automattic/wp-calypso/issues/21665

__The Problem__
Currently the Store UI in Calypso shows the refund button as an option for orders where payment was remitted via PayPal Standard. When one attempts to process the refund, the process fails because no API user/pass/signature has been setup on the payment method. We don't expose the option for setting these values within Store, but they are available in wp-admin.

The proposed solution in this branch is to modify the `method_supports` array that is returned from the `wc/v3/payment_gateways` endpoint. When the the PayPal gateway configuration is missing any/all of the api credentials ( `api_username`, `api_password`, `api_signature` ), the API response will omit `refunds` in the `method_supports` array. If all items are present/not empty strings, the `method_supports` will include `products` and `refunds`. While we don't verify the API credentials are correct, this will at least not signal to API-driven apps like Store on WordPress.com that a refund can be processed via the API for PayPal.

Calypso then uses this `method_supports` data to display either the refund button, or show the manual refund verbiage:

__Refunds with API Creds in Calypso__
![refund-via-paypal](https://user-images.githubusercontent.com/22080/37002599-1c27179e-207f-11e8-9064-bac1f4079eec.png)

__Refund Dialog with NO API Creds__
![manual-refund](https://user-images.githubusercontent.com/22080/37002617-2bc19ef4-207f-11e8-9425-a8e47afbfde6.png)

__To Test__
- Run the `phpunit` test suite to verify the new tests for this are passing
- Use postman and issue GET requests to `/wc/v3/payment_gateways`. First make a request with no api credentials set, verify `method_supports` only shows `products` for PayPal

![method-supports-no-api-keys](https://user-images.githubusercontent.com/22080/37002704-6b4f5408-207f-11e8-8465-6e5e0e7e1326.png)

- Then visit `/wp-admin/admin.php?page=wc-settings&tab=checkout&section=paypal` and add in some jibberish to the API Credentials section

![api-credentials](https://user-images.githubusercontent.com/22080/37002772-a0b66474-207f-11e8-8d25-3536627fa709.png)

- Re-do the GET request to `/wc/v3/payment_gateways` again, and note that `refunds` are now supported:

![refunds-supported](https://user-images.githubusercontent.com/22080/37002815-c638c39a-207f-11e8-809f-af5b027c8fe3.png)
